### PR TITLE
Fix errors when initializing Linear with 0 in_features

### LINKED ISSUE #48152: Allow torch.nn.Linear to have 0 in_features
## 🚀 Feature

Allow [`torch.nn.Linear`](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) to have 0 `in_features`.

Currently, it throws a `ZeroDivisionError` due to the way weights are initialized. See the [`reset_parameters`](https://pytorch.org/docs/stable/_modules/torch/nn/modules/linear.html#Linear) method.

## Motivation

I'm working in a setting where I automatically construct neural networks according to the input features of a given problem. Sometimes, such a problem has no input features at all. For example, in a pure [normal-form game](https://en.wikipedia.org/wiki/Normal-form_game) or [multi-armed bandit problem](https://en.wikipedia.org/wiki/Multi-armed_bandit), there are no observations, and the network simply has to learn a fixed vector of logits. This would correspond to the bias vector of the `torch.nn.Linear`, if it wasn't throwing an error.

Creating a special case for such situations results in cumbersome code.

## Pitch

PyTorch already supports arithmetic with dimensions of size 0:

```python3
import torch
n_features = 0
n_actions = 5
weights = torch.rand(n_actions, n_features)
bias = torch.rand(n_actions)
features = torch.rand(n_features)
logits = weights @ features + bias
```

Fixing this issue would simply require adjusting `reset_parameters`.

cc @albanD @mruberry @jbschlosser

### PR DESCRIPTION
**Pull request by beningodfrey4**
Fixes #48152

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Shouldn't the init methods be fixed instead? This would be less error-brittle, since init functions are also used directly by users, and the same problem would re-appear

**Review by jbschlosser — CHANGES_REQUESTED** [review]
I agree with @vadimkantorov that changes should be made in the init functions to solve this problem more broadly. AFAICT, the only 2 init functions that don't support 0-element tensors are `torch.nn.init.kaiming_normal_` and `torch.nn.init.kaiming_uniform_`, so those should be updated.

Note that the current PR doesn't fully fix the problem- the calculation for the bias initialization is also problematic with 0 `in_features`.

File: torch/nn/init.py [hunk 1]
```diff
@@ -384,6 +384,9 @@ def kaiming_uniform_(tensor, a=0, mode='fan_in', nonlinearity='leaky_relu'):
         >>> w = torch.empty(3, 5)
         >>> nn.init.kaiming_uniform_(w, mode='fan_in', nonlinearity='relu')
     """
+    if 0 not in tensor.shape:
```
Comment (jbschlosser) [inline torch/nn/init.py]: Logic seems reversed here and below. Also maybe checking `tensor.numel() == 0` instead is better?

File: torch/nn/init.py [hunk 2]
```diff
@@ -384,6 +384,9 @@ def kaiming_uniform_(tensor, a=0, mode='fan_in', nonlinearity='leaky_relu'):
...
+        warnings.warn("Initializing zero-dimensional tensors is a no-op")
```
Comment (jbschlosser) [inline torch/nn/init.py:388]: I think this should be "zero-element" tensors instead of "zero-dimensional".

File: torch/nn/modules/linear.py [hunk 3]
```diff
@@ -84,11 +84,14 @@ def __init__(self, in_features: int, out_features: int, bias: bool = True) -> No
...
-        init.kaiming_uniform_(self.weight, a=math.sqrt(5))
-        if self.bias is not None:
-            fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
-            bound = 1 / math.sqrt(fan_in)
-            init.uniform_(self.bias, -bound, bound)
+        if 0 not in self.weight.shape:
+            init.kaiming_uniform_(self.weight, a=math.sqrt(5))
+            if self.bias is not None:
+                fan_in, _ = init._calculate_fan_in_and_fan_out(self.weight)
+                bound = 1 / math.sqrt(fan_in)
```
Comment (jbschlosser) [inline torch/nn/modules/linear.py:91]: Alternatively, this solves the bias issue in a simpler way:
```python
bound = 1 / math.sqrt(fan_in) if fan_in > 0 else 0.
```

**Comment by beningodfrey4** [comment]
I too agree with @vadimkantorov and decided to raise a warning on the 2 init functions if called with a 0-dim tensor cos it's a no-op. 

Regarding the query about bias @jbschlosser, since the initialized weights are the only parameters that lead to non-zero biases, zero-initialization will suffice. Please review https://github.com/pytorch/pytorch/pull/56505/commits/e744bc599ff161447df643b22906527cb87bd9c9 and https://github.com/pytorch/pytorch/pull/56505/commits/9ca8b0733b7bf33d287b8dc682be4e428ffc896f. I presume that there isn't a faster alternative.

**Review by jbschlosser — COMMENTED** [review]
Thanks for your contribution! Would you please also add tests to `test/test_nn.py` ensuring that zero-element tensors can be passed without error to `kaiming_uniform_` / `kaiming_normal_`? A good place is next to the other `kaiming_*` tests: https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L10598-L10663

File: torch/nn/init.py (hunk 2 above)
Comment (beningodfrey4) [reply torch/nn/init.py]: Agreed, fixed in https://github.com/pytorch/pytorch/pull/56505/commits/9f86953b32fea04694eccf2b0d711820d05a0fd8

File: torch/nn/modules/linear.py (hunk 3 above)
Comment (beningodfrey4) [reply torch/nn/modules/linear.py]: Much better. Fixed in https://github.com/pytorch/pytorch/pull/56505/commits/9f86953b32fea04694eccf2b0d711820d05a0fd8

**Review by jbschlosser — APPROVED** [review]
LGTM! Once the CI checks are green :)

**Comment by beningodfrey4** [comment]
Woking on them. Thank you very much for your guidance. If this goes through, it will be my first OSS contribution :)

File: test/test_nn.py [hunk 4]
```diff
@@ -10607,6 +10607,16 @@ def test_kaiming_normal_errors_on_inputs_smaller_than_2d(self):
                 tensor = self._create_random_nd_tensor(dims, size_min=1, size_max=1)
                 init.kaiming_normal_(tensor)
 
+    def test_kaiming_uniform_warning_on_0element_tensor(self):
+        tensor = self.empty(0, 1)
```
Comment (jbschlosser) [inline test/test_nn.py:10611]: I think you want `torch.empty` here and below instead of `self.empty`.

Comment (beningodfrey4) [reply test/test_nn.py]: Sorry for the delay. Fixed in https://github.com/pytorch/pytorch/pull/56505/commits/baf05333ed74e4581157a483d4800c3613e36ce1.

**Comment by jbschlosser** [comment]
FYI the `pytorch_xla_linux_bionic_py3_6_clang9_build` failure is unrelated.
